packages/@aws-cdk/aws-amplify/lib/app.ts

```diff
@@ -0,0 +1,347 @@
+
+  private readonly environmentVariablesResolver: EnvironmentVariablesResolver = new EnvironmentVariablesResolver();
+
+  private readonly customRulesResolver: CustomRulesResolver = new CustomRulesResolver();
+
+  constructor(scope: Construct, id: string, props: AppProps) {
+    super(scope, id, {
+      physicalName: props.appName
+    });
+
+    if (props.basicAuthConfig) {
+      this.addBasicAuth(props.basicAuthConfig.username, props.basicAuthConfig.password);
+    }
+
+    if (props.environmentVariables && props.environmentVariables.length > 0) {
+      this.environmentVariablesResolver.addEnvironmentVariables(...props.environmentVariables);
+    }
+
+    let roleArn: string | undefined;
+
+    if (props.serviceRole) {
+      roleArn = props.serviceRole.roleArn;
+    }
+
+    if (props.buildSpec && !props.serviceRole) {
+      const role = new LazyRole(scope, 'lazy-role', {
+        assumedBy: new ServicePrincipal('amplify.amazonaws.com')
+      });
+
+      roleArn = role.roleArn;
+    }
+
+    const resource = new CfnApp(this, 'App', {
+      accessToken: props.accessToken,
+      basicAuthConfig: this.basicAuthResolver,
+      buildSpec: (props.buildSpec) ? props.buildSpec.toBuildSpec() : undefined,
+      customRules: this.customRulesResolver,
+      description: props.description,
+      environmentVariables: this.environmentVariablesResolver,
+      iamServiceRole: roleArn,
+      name: props.appName,
+      oauthToken: (props.oauthToken && props.oauthToken.toString()) ? props.oauthToken.toString() : undefined,
+      repository: props.repository,
+      tags: props.tags
+    });
+
+    this.appArn = resource.attrArn,
+    this.appDefaultDomain = resource.attrDefaultDomain;
+    this.appId = resource.attrAppId;
+    this.appName = resource.attrAppName;
+
+    this.resource = resource;
+  }
+
+  /**
+   * Add Service Role
+   *
+   * @param role
+   */
+  public addServiceRole(role: IRole) {
+    this.resource.iamServiceRole = role.roleArn;
+  }
+
+  /**
+   * Add Basic Auth on App
+   *
+   * @param username
+   * @param password
+   */
+  public addBasicAuth(username: string, password: string) {
+    this.basicAuthResolver.basicAuthConfig({
+      enableBasicAuth: true,
+      password,
+      username
+    });
+  }
+
+  /**
+   * Add Environment Variable to App
+   *
+   * @param name
+   * @param value
+   */
+  public addEnvironmentVariable(name: string, value: string) {
+    this.environmentVariablesResolver.addEnvironmentVariable(name, value);
+  }
+
+  /**
+   * Add custom rule to app
+   *
+   * @param source
+   * @param target
+   * @param status
+   * @param condition
+   */
+  public addCustomRule(source: string, target: string, status?: string, condition?: string) {
+    this.customRulesResolver.addCustomRule(source, target, status, condition);
+  }
+
+  /**
+   * Add domain to app
+   *
+   * @param id
+   * @param props
+   */
+  public addDomain(id: string, props: DomainBaseProps): Domain {
```

Comment:
The "props" terminology is only used for constructors.

* Rename `XxxBaseProps` to `XxxOptions` (in this case `DomainOptions`)
* Rename `props` to `options`.